gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/writer/HiveMetadataWriter.java

```diff
@@ -0,0 +1,287 @@
+      for (HiveSpec spec : specs) {
+        if (spec.getTable().getDbName().equals(dbName) && spec.getTable().getTableName().equals(tableName)) {
+             if (spec.getPartition().isPresent()) {
+              deRegisterPartitionHelper(dbName, tableName, spec);
+             }
+        }
+      }
+    }
+    //TODO: De-register table if table location does not exist
+
+  }
+
+  protected void deRegisterPartitionHelper(String dbName, String tableName, HiveSpec spec) throws IOException {
+    hiveRegister.dropPartitionIfExists(dbName, tableName,
+        spec.getTable().getPartitionKeys(), spec.getPartition().get().getValues());
+  }
+
+  public void addFiles(GobblinMetadataChangeEvent gmce, Map<String, Collection<HiveSpec>> newSpecsMap, String dbName,
+      String tableName, String topicName) throws SchemaRegistryException {
+    String tableKey = tableNameJoiner.join(dbName, tableName);
+    for (Collection<HiveSpec> specs : newSpecsMap.values()) {
+      for (HiveSpec spec : specs) {
+        if (spec.getTable().getDbName().equals(dbName) && spec.getTable().getTableName().equals(tableName)) {
+          List<String> partitionValue =
+              spec.getPartition().isPresent() ? spec.getPartition().get().getValues() : Lists.newArrayList();
+          Cache<List<String>, HiveSpec> hiveSpecCache = specMaps.computeIfAbsent(tableKey,
+              s -> CacheBuilder.newBuilder()
+                  .expireAfterAccess(state.getPropAsInt(MetadataWriter.CACHE_EXPIRING_TIME,
+                      MetadataWriter.DEFAULT_CACHE_EXPIRING_TIME), TimeUnit.HOURS)
+                  .build());
+          HiveSpec existedSpec = hiveSpecCache.getIfPresent(partitionValue);
+          if (existedSpec != null) {
+            //if existedSpec is not null, it means we already registered this partition, so check whether we need to update the table/partition
+            if (gmce.getSchemaSource() == SchemaSource.SCHEMAREGISTRY) {
+              // Only in this case, we need to update the schema to the latest schema in schemaRegistry
+              schemaUpdateHelper(spec, topicName, tableKey);
+            }
+            if ((this.hiveRegister.needToUpdateTable(existedSpec.getTable(), spec.getTable())) ||
+                (spec.getPartition().isPresent() && this.hiveRegister.needToUpdatePartition(
+                    existedSpec.getPartition().get(), spec.getPartition().get()))) {
+              registerSpec(tableKey, partitionValue, spec, hiveSpecCache);
+            }
+          } else {
+            registerSpec(tableKey, partitionValue, spec, hiveSpecCache);
+          }
+        }
+      }
+    }
+  }
+
+  private void registerSpec(String tableKey, List<String> partitionValue, HiveSpec spec,
+      Cache<List<String>, HiveSpec> hiveSpecCache) {
+    HashMap<List<String>, ListenableFuture<Void>> executionMap =
+        this.currentExecutionMap.computeIfAbsent(tableKey, s -> new HashMap<>());
+    if (executionMap.containsKey(partitionValue)) {
```

Comment:
I don't quite understand this part. Say we have an old `spec` to register ( meaning the executionMap contains it). We wait on the call to finish by calling future.get() in line 226, and then call it again in line 232 ?  

Also, when we find a branch new spec, we simply put the future object in the executionMap but why we want to ensure it finished by calling future.get() next time we see it ? Isn't this barrier being put in the flush method ?

Reply:
When we call registerSpec, it means we see a brand new spec: 
1. It's a new partition and we haven't see it before, in this case, we will skip 226 and direct register in 232. 
2. We have register this partition before and we see a updated spec for this partition (i.e. schema change), in this case, we want to make sure old one and the new one are registered in order so that there won't be situation that we try to register old partition spec A and it's pending in the queue, and then we submit another task to register new spec B, but during execution, B goes first and finally the partition in metaStore is A(the old one)
Since there should not be too many update in same partition during a flush interval, so this should not affect performance so much

Reply:
I see. It makes more sense, thanks !